### PR TITLE
pl061 gpio v1

### PR DESCRIPTION
**Pull request by vchong**
Signed-off-by: Victor Chong <victor.chong@linaro.org>

### CODE REVIEW

**Comment by vchong** [comment]
This is ported directly from ARM-TF with the permission of the original patch author.

File: core/drivers/pl061_gpio.c [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+# define PLAT_PL061_MAX_GPIOS	32
+#endif	/* PLAT_PL061_MAX_GPIOS */
+
+#define MAX_GPIO_DEVICES	((PLAT_PL061_MAX_GPIOS +		\
+					(GPIOS_PER_PL061 - 1)) / GPIOS_PER_PL061)
+
+#define PL061_GPIO_DIR		0x400
+
+#define GPIOS_PER_PL061		8
+#define BIT(nr)			(1UL << (nr))
```
Comment (jenswi-linaro) [inline core/drivers/pl061_gpio.c]: This define could be useful elsewhere too, please move it to `lib/libutils/ext/include/util.h`

File: core/include/gpio.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __GPIO_H__
+#define __GPIO_H__
+
+#define GPIO_DIR_OUT		0
```
Comment (jenswi-linaro) [inline core/include/gpio.h]: The interface would be more clear with an enum instead.

File: core/include/gpio.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+#define GPIO_DIR_IN		1
+
+#define GPIO_LEVEL_LOW		0
```
Comment (jenswi-linaro) [inline core/include/gpio.h]: The interface would be more clear with an enum instead.

File: core/include/gpio.h [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+#define GPIO_LEVEL_HIGH		1
+
+typedef struct gpio_ops {
```
Comment (jenswi-linaro) [inline core/include/gpio.h]: Please drop the typedef

File: core/drivers/pl061_gpio.c (hunk 1 above)
Comment (vchong) [inline core/drivers/pl061_gpio.c]: Ok

File: core/include/gpio.h (hunk 2 above)
Comment (vchong) [inline core/include/gpio.h]: Ok. Should I do it for GPIO_LEVEL_* too?

Comment (vchong) [inline core/include/gpio.h]: Oops nvm. See comment below now.

File: core/include/gpio.h (hunk 3 above)
Comment (vchong) [inline core/include/gpio.h]: Ok

File: core/include/gpio.h (hunk 4 above)
Comment (vchong) [inline core/include/gpio.h]: It won't build if just typedef is removed. Do you mean drop typedef and replace `gpio_ops_t` with `struct gpio_ops` everywhere it's being used? If yes, why?

Comment (jenswi-linaro) [inline core/include/gpio.h]: Yes, we're using Linux kernel style.
https://github.com/linaro-swg/linux/blob/optee/Documentation/CodingStyle#L273

Comment (vchong) [inline core/include/gpio.h]: Ok, got it. Will change accordingly. Thanks!

**Comment by vchong** [comment]
Rebased and review comments addressed.

File: core/include/gpio.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	GPIO_DIR_IN
+};
+
+enum {
+	GPIO_LEVEL_LOW,
+	GPIO_LEVEL_HIGH
+};
+
+struct gpio_ops {
+	int (*get_direction)(int gpio);
```
Comment (jenswi-linaro) [inline core/include/gpio.h]: Please make these functions take enums where applicable instead of int.

File: core/include/drivers/pl061_gpio.h [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __PL061_GPIO_H__
+#define __PL061_GPIO_H__
+
+#include <gpio.h>
+
+void pl061_gpio_register(vaddr_t base_addr, int gpio_dev);
```
Comment (jenswi-linaro) [inline core/include/drivers/pl061_gpio.h]: I'd rather use `base_addr` to identify a gpio device than just another id that we still need to keep track of.
Another advantage is that the driver doesn't have to try to estimate how many gpio devices we need.

File: core/include/gpio.h (hunk 5 above)
Comment (vchong) [inline core/include/gpio.h]: Ok

File: core/include/drivers/pl061_gpio.h (hunk 6 above)
Comment (vchong) [inline core/include/drivers/pl061_gpio.h]: Assuming that by 'id' you mean the `gpio_dev` arg?

The id actually simplifies the rest of the code a lot. We can use a simplified way, i.e. just use `gpio / GPIOS_PER_PL061` to calculate the index into array of registered base addresses to find the right base address to use. It does put the burden on the platform (code) to register/index the gpio devices properly, but it should be easy to do, i.e. the first device is 0, 2nd device is 1, 3rd dev 2, and so on so forth.

Without the id, and by just using `base address`, we would still need some sort of indexed array anyway, usually within a `struct` somewhere internally within the pl061 driver code, to keep track of all the available devices, and repeated loopings through the array to calculate the proper base address required for every get_ and set_ calls based on the gpio pin/number provided.

This was actually how it was done originally [1][2] until ARM proposed the simplification [3].

[1] https://github.com/96boards-hikey/arm-trusted-firmware/blob/hikey/drivers/arm/gpio/gpio.c#L118-L132
[2] https://github.com/ARM-software/arm-trusted-firmware/pull/510/commits/fe8880cc0771539fb7112f838bd93fb312af2be7#diff-ee806812453e4998cf59d65d6ceab1afR150
[3] https://github.com/hzhuang1/arm-trusted-firmware/commit/b688de6c94485847b839f6225eb36c128e66d3a1#diff-ee806812453e4998cf59d65d6ceab1afR137

File: core/include/gpio.h (hunk 5 above)
Comment (vchong) [inline core/include/gpio.h]: Is this what you mean? It looks somewhat more complicated than before, so thought I'd better check first.
https://github.com/vchong/optee_os/compare/gpio...vchong:gpio3?expand=1

Comment (jenswi-linaro) [inline core/include/gpio.h]: Yes

File: core/include/drivers/pl061_gpio.h (hunk 6 above)
Comment (vchong) [inline core/include/drivers/pl061_gpio.h]: Comment above edited for better clarity.

File: core/drivers/pl061_gpio.c [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+#include <io.h>
+#include <util.h>
+#include <drivers/pl061_gpio.h>
+
+#ifndef PLAT_PL061_MAX_GPIOS
+# define PLAT_PL061_MAX_GPIOS	32
+#endif	/* PLAT_PL061_MAX_GPIOS */
+
+#define MAX_GPIO_DEVICES	((PLAT_PL061_MAX_GPIOS +		\
+					(GPIOS_PER_PL061 - 1)) / GPIOS_PER_PL061)
```
Comment (pascal-brand-st-dev) [inline core/drivers/pl061_gpio.c]: Line over 80 chars suggested by Travis

File: core/drivers/sub.mk [hunk 8]
```diff
@@ -1,5 +1,7 @@
 srcs-$(CFG_PL011) += pl011.c
 srcs-$(CFG_GIC) += gic.c
+srcs-$(CFG_GPIO) += gpio.c
+srcs-$(CFG_PL061) += pl061_gpio.c
```
Comment (pascal-brand-st-dev) [inline core/drivers/sub.mk]: Should we add

    #ifdef CFG_PL061
    $(call force,CFG_GPIO,y,required by CFG_PL061) 
    #endif

**Comment by pascal-brand-st-dev** [comment]
Add a `Suggested-By` tag with the original author

File: core/drivers/pl061_gpio.c (hunk 7 above)
Comment (vchong) [inline core/drivers/pl061_gpio.c]: Ok. Will reduce.

File: core/drivers/sub.mk (hunk 8 above)
Comment (vchong) [inline core/drivers/sub.mk]: Yes good idea. Thanks.

**Comment by vchong** [comment]
Ah yes, I'd meant to do that but sort of neglected it. Thanks for the reminder!

File: core/include/drivers/pl061_gpio.h (hunk 6 above)
Comment (jenswi-linaro) [inline core/include/drivers/pl061_gpio.h]: OK, I see the point with registering the base addresses now. I didn't realize that `gpio_dev` wasn't the same thing as `gpio`. Perhaps `gpio` should be renamed to `gpio_pin` for clarity.

Comment (vchong) [inline core/include/drivers/pl061_gpio.h]: Ok, I guess it can be confusing. I will change `gpio` to `gpio_pin` in v2. Thanks!

**Comment by vchong** [comment]
Replaced by #787.

**Pull request closed without merging**